setup.py

```diff
@@ -35,14 +35,93 @@ def is_conda_interpreter():
     """
     return 'conda' in sys.version or 'Continuum' in sys.version
 
+def is_macOS():
+    return 'darwin' in sys.platform
+
+
+# HOMEBREW_SAMPLE = """$ brew ls --verbose libomp
+# /opt/homebrew/Cellar/libomp/15.0.7/INSTALL_RECEIPT.json
+# /opt/homebrew/Cellar/libomp/15.0.7/.brew/libomp.rb
+# /opt/homebrew/Cellar/libomp/15.0.7/include/ompt.h
+# /opt/homebrew/Cellar/libomp/15.0.7/include/omp.h
+# /opt/homebrew/Cellar/libomp/15.0.7/include/omp-tools.h
+# /opt/homebrew/Cellar/libomp/15.0.7/lib/libomp.dylib
+# /opt/homebrew/Cellar/libomp/15.0.7/lib/libomp.a
+# """
+#
+# MACPORTS_SAMPLE = """$ port contents libomp
+# Port libomp contains:
+#   /opt/local/include/libomp/omp-tools.h
+#   /opt/local/include/libomp/omp.h
+#   /opt/local/include/libomp/ompt.h
+#   /opt/local/lib/libomp/libgomp.dylib
+#   /opt/local/lib/libomp/libiomp5.dylib
+#   /opt/local/lib/libomp/libomp.dylib
+#   /opt/local/share/doc/libomp/LICENSE.TXT
+#   /opt/local/share/doc/libomp/README.txt
+# """
+
+def macOS_list_brew_port(cmd):
+    from subprocess import run
+    lib = run(cmd + " | awk '/\/lib\// { print $1; }' | xargs -n1 dirname | uniq",
+              shell=True, check=False, capture_output=True)
+    inc = run(cmd + " | awk '/\/include\// { print $1; }' | xargs -n1 dirname | uniq",
+              shell=True, check=False, capture_output=True)
+    if lib.returncode == 0 and inc.returncode == 0:
+        return inc.stdout.decode("UTF-8").strip(), lib.stdout.decode("UTF-8").strip()
+    return None, None
+
+MACOS_OMP = None
+
+def macOS_libomp_known_locations():
+    """ Find common paths for libomp installation on macOS
+        return flags for compile and link as lists
+        e.g. (['-I/opt/local/include/libomp'], ['-L/opt/local/lib/libomp'])
+    """
+    if not is_macOS():
+        return [], []
+    global MACOS_OMP
+    if MACOS_OMP:
+        return MACOS_OMP
+    for cmd in ["brew ls --verbose libomp", "port contents libomp"]:
+        inc, lib = macOS_list_brew_port(cmd)
+        if inc and lib:
+            MACOS_OMP = [f"-I{inc}"], [f"-L{lib}"]
+            return MACOS_OMP
```

Comment:
Why do we need the globals here? I see this function is called twice, right? Perhaps we could move the `OMP_TAB` logic to inside the `build_extensions` method and reuse some of these conditions. It could make it more clear (if factored a lot) which conditions apply where.

Reply:
Integrating this into the build_ext_subclass sounds like a good approach for this; the intent is strictly to make sure we're only shelling out to probe the minimum number of times. Could also potentially be done with a decorator. I started thinking about making this into its own OpenMP-options-probing class that provides all the table information as calls or attributes, but decided to stop and share while I had something useful enough.

Reply:
I'm not sure how clean the code would be, but a class that does all the probing on `__init__` (or on `.probe()` call) and then has properties for all of the information that may be needed for compilation could be nice. Like `.openmp_exists`, `.openmp_cflags`, `.openmp_ldflags`, etc. You could even more a lot of the Compiler based decisions in there.

Then if someone wants to add functionality for other compilers or OpenMP-like libraries there is a pretty clear interface to fit into.